lib/web_ui/lib/src/engine/scene_view.dart

```diff
@@ -0,0 +1,235 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:ui/src/engine.dart';
+import 'package:ui/ui.dart' as ui;
+
+const String kCanvasContainerTag = 'flt-canvas-container';
+
+// This is an interface that renders a `ScenePicture` as a `DomImageBitmap`.
+// It is optionally asynchronous. It is required for the `EngineSceneView` to
+// composite pictures into the canvases in the DOM tree it builds.
+abstract class PictureRenderer {
+  FutureOr<DomImageBitmap> renderPicture(ScenePicture picture);
+}
+
+// This class builds a DOM tree that composites an `EngineScene`.
+class EngineSceneView {
+  factory EngineSceneView(PictureRenderer pictureRenderer) {
+    final DomElement sceneElement = createDomElement('flt-scene');
+    return EngineSceneView._(pictureRenderer, sceneElement);
+  }
+
+  EngineSceneView._(this.pictureRenderer, this.sceneElement);
+
+  final PictureRenderer pictureRenderer;
+  final DomElement sceneElement;
+
+  List<SliceContainer> containers = <SliceContainer>[];
+
+  int queuedRenders = 0;
+  static const int kMaxQueuedRenders = 3;
+
+  Future<void> renderScene(EngineScene scene) async {
+    if (queuedRenders >= kMaxQueuedRenders) {
+      return;
+    }
+    queuedRenders += 1;
+
+    scene.beginRender();
+    final List<LayerSlice> slices = scene.rootLayer.slices;
+    final Iterable<Future<DomImageBitmap?>> renderFutures = slices.map(
```

Comment:
nit: might be cleaner to use whereType to filter out the PlatformViewSlices rather than make a list of nullable futures

Reply:
The reason I did it this way is so that I can extract them by index when I'm iterating the slices themselves. If we use `whereType` there is no simple way to correlate the image bitmap with the slice it was rendered from.